### DISCUSSION TITLE
Gatweay Port Hangs - How to health check

### DISCUSSION THREAD

**Discussion by mpelteshki** [gnzsnz/ib-gateway-docker#124] [Q&A] [answered]
## Describe the bug

The docker image runs well, the terminal logs no errors as far as I have seen (I will attach a log just in case). But when I do a "curl -X GET "localhost:$port", the command hangs. I am sure I am using the right port as when I mistake the live for the paper port (or vice versa), the command returns an error. 

Note: I am using a custom image, but it is based on the latest stable with the only difference being that I installed curl for healthcheck purposes. I may be missing something really simple, perhaps something firewall related? I haven't touched firewall settings. But I just can't figure it out. Any help would be appreciated.

docker-compose.yaml:
<img width="762" alt="image" src="https://github.com/gnzsnz/ib-gateway-docker/assets/39273158/cc5b84ee-62f8-4c6c-917e-8f17cfe4cafa">

.env:
<img width="762" alt="image" src="https://github.com/gnzsnz/ib-gateway-docker/assets/39273158/0e3624d3-b9ac-4bf4-82a8-4a46236c9375">
(username and password are filled in when running, the rest are as is)

I build the project with 'docker compose up'

## Expected

The second package is a python app that calls the TWS API. I need the ib-gateway to make the api available so the healthcheck passes and the app connects to the API.

## Container logs
[logs.txt](https://github.com/user-attachments/files/15865823/logs.txt)## Versions

Please complete the following information:

- OS: macOS 14
- Image Tag: stable

## Additional context

What have you tried and failed.
I have tried stable/latest branches
I have tried accessing the port from outside of the container, same result.
I have tried accessing the port from inside of the container, same result.
I have tried using a paper account instead of a live one (ports adjusted), same result.
I have tried using 127.0.0.1:4001:4003 instead of 4001:4003.

**Comment by gnzsnz** [comment, marked as answer]
Hi,

not shure what's your expectation, why would `"curl -X GET "localhost:$port/v1/api/one/user"` work, this is not a RESTfull API.

I just did a quick test on my running TWS instance and ports are working.

![image](https://github.com/gnzsnz/ib-gateway-docker/assets/8376642/08dd5db6-fcc7-4ae2-b3b6-ba196f8c6ca4)

if you need a healthcheck(which just is telling you that the port is open, but not that is actually working) you can use that

this is how it would look like on ib-gateway with the port down (ibgateway is not running)

![image](https://github.com/gnzsnz/ib-gateway-docker/assets/8376642/e90be4a3-753d-4161-8e53-4e14e6a4ec65)

something like this to have it working as a docker healthcheck

```dockerfile
HEALTHCHECK --interval=30m --timeout=15s --start-period=10s \
  CMD timeout 1 bash -c '</dev/tcp/0.0.0.0/4001 && echo "TWS API up" || echo "TWS API down"' || echo "Connection timeout"
```

take into account that any disconnection to IBKR server farm will not be reported by this health check. I use ib_async and disconnectedEvent to keep TWS api connectivity.

hope this helps

**Comment by mpelteshki** [comment]
Hm, yeah thanks for the healthcheck, I could see it being useful. And thanks for the answer, I figured it out. I wasn't managing my connection IDs correctly.
